pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java

```diff
@@ -0,0 +1,302 @@
+    // 3. Sort all the segments to be used during assignment in ascending order of replicas
+    // 4. For each segment to be used during assignment, pick a random replica, weighted by the number of segments
+    //    assigned to each replica.
+
+    // 1. Gather all segments and group them by Kafka partition, sorted by sequence number
+    Map<String, SortedSet<String>> sortedSegmentsByKafkaPartition = new HashMap<String, SortedSet<String>>();
+    for (String helixPartitionName : externalView.getPartitionSet()) {
+      // Ignore segments that are not low level consumer segments
+      if (!SegmentNameBuilder.Realtime.isRealtimeV2Name(helixPartitionName)) {
+        continue;
+      }
+
+      String kafkaPartitionName = SegmentNameBuilder.Realtime.extractPartitionRange(helixPartitionName);
+      SortedSet<String> segmentsForPartition = sortedSegmentsByKafkaPartition.get(kafkaPartitionName);
+
+      // Create sorted set if necessary
+      if (segmentsForPartition == null) {
+        segmentsForPartition = new TreeSet<String>(new Comparator<String>() {
+          @Override
+          public int compare(String firstSegment, String secondSegment) {
+            // Sort based on sequence number, falling back on string comparison in case there is an exception
+            try {
+              int firstSegmentSequenceNumber =
+                  Integer.parseInt(SegmentNameBuilder.Realtime.extractSequenceNumber(firstSegment));
+              int secondSegmentSequenceNumber =
+                  Integer.parseInt(SegmentNameBuilder.Realtime.extractSequenceNumber(secondSegment));
+              return Integer.compare(firstSegmentSequenceNumber, secondSegmentSequenceNumber);
+            } catch (NumberFormatException e) {
+              LOGGER.warn("Caught number format exception while comparing segments {} and {}", firstSegment,
+                  secondSegment, e);
+              return firstSegment.compareTo(secondSegment);
+            }
+          }
+        });
+
+        sortedSegmentsByKafkaPartition.put(kafkaPartitionName, segmentsForPartition);
+      }
+
+      segmentsForPartition.add(helixPartitionName);
+    }
+
+    // 2. Ensure that for each partition, we have at most one partition in consuming state
+    Map<String, String> lastSegmentInConsumingStateByKafkaPartition = new HashMap<String, String>();
+    for (String kafkaPartition : sortedSegmentsByKafkaPartition.keySet()) {
+      SortedSet<String> sortedSegmentsForKafkaPartition = sortedSegmentsByKafkaPartition.get(kafkaPartition);
+      String lastSegment = sortedSegmentsForKafkaPartition.last();
+
+      // Only keep the segment if all replicas have it in CONSUMING state
+      Map<String, String> helixPartitionState = externalView.getStateMap(lastSegment);
+      for (String externalViewState : helixPartitionState.values()) {
+        // Ignore ERROR state
+        if (externalViewState.equalsIgnoreCase(
```

Comment:
No, only the last segment is one for which the consuming state is valid. This is used to store whether the last segment is a segment for which the `CONSUMING` state is valid or not. Any other segment should be in `ONLINE` state to be routable.